README.md
# MMM-horoscope
Horoscope Module for MagicMirror² v.0.1.0


This is a module for [MagicMirror²](https://github.com/MichMich/MagicMirror) that adds daily horoscope pulled from the [Yahoo Style Horoscopes](https://www.yahoo.com/style/horoscope/). API endpoint used for this module is not officially supported and may break at any time.

![screenshot](screenshot.png)

## Dependencies

npm modules (should already be available with your installation):

- request
- moment

## Using the module

To use this module, add it to the modules array in the `config/config.js` file:

````javascript
modules: [
  {
    module: 'MMM-horoscope',
    position: 'top_right',  // This can be any of the regions.
    config: {
      // See 'Configuration options' for additional options and more information.
      sign: 'aries', // Zodiac sign
    }
  }
]
````

## Configuration options

<table width="100%">
  <!-- why, markdown... -->
  <thead>
    <tr>
      <th>Option</th>      
      <th width="100%">Description</th>
    </tr>
  <thead>
  <tbody>
    <tr>
      <td><code>sign</code></td>      
      <td>Zodiac sign for daily horoscope. (String)<br>
        <br>This value is <b>REQUIRED</b><br>
        <br><b>Possible values:</b> <code>aries</code>, <code>taurus</code>, <code>gemini</code>, <code>cancer</code>, <code>leo</code>, <code>virgo</code>, <code>libra</code>, <code>scorpio</code>, <code>sagittarius</code>, <code>capricorn</code>,<code>aquarius</code>,<code>pisces</code>
        <br><b>Default value:</b> <code>aries</code>
      </td>
    </tr>
    <tr>
      <td><code>maxWidth</code></td>
      <td>Overall width of the module to help you fit it in your layout (String CSS value)<br>
        <br><b>Possible values can be specified in:</b> <code>px</code>, <code>%</code>, <code>em</code> etc.
        <br><b>Default value:</b> <code>400px</code>
      </td>
    </tr>
    <tr>
      <td><code>updateInterval</code></td>
      <td>How often does the content needs to be fetched? (Milliseconds)<br>
      	<br>Since we are looking at daily horoscope data doesn't change much it is recommended not to pull it very often. Once an hour should be more than enough.<br>
        <br><b>Possible values:</b> <code>1000</code>-<code>86400000</code>
        <br><b>Default value:</b> <code>3600000</code> (1 hour)
      </td>
    </tr>
    <tr>
      <td><code>timeShift</code></td>
      <td>Shifting time to display next or previous day horoscope. (Milliseconds)<br>
      	<br>This is done for conveniece of displaying next day horoscope in the evening. For example you'd like to start displaying next day horoscope after 7pm, in that case you would need to set <code>timeShift</code> value to <code>18000000</code>. Math works out to <code>24 - 19 = 5</code>, converting 5 hours into milliseconds <code>5 * 60 * 60 * 1000 = 18000000</code><br>
        <br><b>Possible values:</b> <code>0</code>-<code>86400000</code>
        <br><b>Default value:</b> <code>18000000</code> (5 hours)
      </td>
    </tr>
     <tr>
      <td><code>useTextIcon</code></td>
      <td>Display Zodiac sign using text or image for the module output (Boolean)<br>
      	<br>See screenshot: left side is <code>useTextIcon=true</code>, right side is <code>useTextIcon=false</code>. CSS styles are using DejaVu Sans font which should be present on RasPi, other platforms may need CSS stylesheet adjustments.<br>
        <br><b>Possible values:</b> <code>true</code> or <code>false</code>
        <br><b>Default value:</b>  <code>true</code>
      </td>
    </tr>
    <tr>
      <td><code>initialLoadDelay</code></td>
      <td>The initial delay before loading. (Milliseconds)<br>
      <br>Largely Unused. If you have multiple modules that use the same API endpoint, you might want to delay one of the requests.<br>
        <br><b>Possible values:</b> <code>1000</code> - <code>5000</code>
        <br><b>Default value:</b>  <code>0</code>
      </td>
    </tr>
    <tr>
      <td><code>animationSpeed</code></td>
      <td>Speed of the update animation. (Milliseconds)<br>
        <br><b>Possible values:</b><code>0</code> - <code>5000</code>
        <br><b>Default value:</b> <code>2000</code> (2 seconds)
      </td>
    </tr>
  </tbody>
</table>


## Support

This is my very first MM module and I probably missed something obvious. Please get in touch with me if you have a suggestion or feature request. Feel free to file a [GitHub issue](https://github.com/morozgrafix/MMM-horoscope/issues)

CHANGELOG.md
# MMM-horoscope Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.0] - 2017-01-05

# Initial public release

node_helper.js
/* MMM-horoscope
 * Node Helper
 *
 * By morozgrafix https://github.com/morozgrafix/MMM-horoscope
 *
 * License: MIT
 *
 * Based on https://github.com/fewieden/MMM-soccer/blob/master/node_helper.js
 *
 */

const request = require("request");
const NodeHelper = require("node_helper");

module.exports = NodeHelper.create({
	// subclass start method
	start: function(){
		console.log("Starting NodeHelper for " + this.name + "module.");
	},

	// subclass socketNotificationReceived method
	socketNotificationReceived: function(notification, payload) {
		if (notification === "GET_HOROSCOPE_DATA") {
			var query = "{\"sign\":\"" + payload.sign.substring(0,3) + "\",\"frequency\":\"daily\",\"date\":\"" + payload.date + "\",\"week\":\"\",\"month\":\"\"}";
			var options = {
				url: "https://www.yahoo.com/_td/api/resource/horoscope.astro;query=" + encodeURIComponent(query) + ";site=horoscope"
			};
			this.getData(options);
		}
	},

	// get data from URL and broadcast it to MagicMirror module if everyting is OK
	getData: function(options) {
		request(options, (error, response, body) => {
			if (response.statusCode === 200) {
				this.sendSocketNotification("HOROSCOPE_DATA", JSON.parse(body));
			} else {
				this.sendSocketNotification("HOROSCOPE_DATA");
				console.log("Error getting Horoscope data. Response:" + JSON.stringify(response));
			}
		});
	}
});


MMM-horoscope.js
Module.register("MMM-horoscope",{
	defaults: {
		sign: "aries",
		maxWidth: "400px", // maximum width of the module in px, %, em
		updateInterval: 1 * 60 * 60 * 1000, // updates every hour
		timeShift: 5 * 60 * 60 * 1000, // shift clock in milliseconds to start showing next day horoscope at 7pm (24 - 19 = 5)
		useTextIcon: true,
		initialLoadDelay: 0,
		animationSpeed: 2000,
		zodiacTable: {
			"aries": {
				"signId": "ari",
				"range": "3/21-4/19",
				"unicodeChar": "&#9800;"
			},
			"taurus": {
				"signId": "tau",
				"range": "4/20-5/20",
				"unicodeChar": "&#9801;"
			},
			"gemini": {
				"signId": "gem",
				"range": "5/21-6/21",
				"unicodeChar": "&#9802;"
			},
			"cancer": {
				"signId": "can",
				"range": "6/22-7/22",
				"unicodeChar": "&#9803;"
			},
			"leo": {
				"signId": "leo",
				"range": "7/23-8/22",
				"unicodeChar": "&#9804;"
			},
			"virgo": {
				"signId": "vir",
				"range": "8/23-9/22",
				"unicodeChar": "&#9805;"
			},
			"libra": {
				"signId": "lib",
				"range": "9/23-10/22",
				"unicodeChar": "&#9806;"
			},
			"scorpio": {
				"signId": "sco",
				"range": "10/23-11/21",
				"unicodeChar": "&#9807;"
			},
			"sagittarius": {
				"signId": "sag",
				"range": "11/22-12/21",
				"unicodeChar": "&#9808;"
			},
			"capricorn": {
				"signId": "cap",
				"range": "12/22-1/19",
				"unicodeChar": "&#9809;"
			},
			"aquarius": {
				"signId": "aqu",
				"range": "1/20-2/18",
				"unicodeChar": "&#9810;"
			},
			"pisces": {
				"signId": "pis",
				"range": "2/19-3/20",
				"unicodeChar": "&#9811;"
			}
		},
		debug: false
	},

	getStyles: function() {
		return ["MMM-horoscope.css"];
	},

	getScripts: function() {
		return [
			"moment.js"
		];
	},

	start: function() {
		Log.info("Starting module: " + this.name);
		if (this.config.debug) {
			this.config.updateInterval = 60 * 1000; // update very 1 minute for debug
		}
		// just in case someone puts mixed case in their config files
		this.config.sign = this.config.sign.toLowerCase();
		this.sign = null;
		this.signText = null;
		this.horoscopeText = null;
		this.horoscpeDate = null;
		this.scheduleUpdate(this.config.initialLoadDelay);
	},

	updateHoroscope: function() {
		this.date = new Date();
		this.sendSocketNotification("GET_HOROSCOPE_DATA", { sign: this.config.sign, date: moment(this.date).add(this.config.timeShift, "milliseconds").format("YYYYMMDD") });
	},


	// Subclass socketNotificationReceived method.
	socketNotificationReceived: function(notification, payload){
		if(notification === "HOROSCOPE_DATA" && payload != null){
			this.horoscopeData = payload[0];
			this.processHoroscope(this.horoscopeData);
		} else {
			this.processHoroscopeError("Unable to get horoscope from API. Please check the logs.");
		}
	},

	processHoroscope: function(data) {
		const regex = /<a.*\/a>/g;
		const subst = "";
		this.sign = this.config.zodiacTable[this.config.sign]["unicodeChar"];
		this.signText = data.sign;
		this.horoscopeText = data.overview.replace(regex, subst).trim();
		this.horoscopeDate = data.frequencyValue
		this.loaded = true;
		this.updateDom(this.config.animationSpeed);
		this.scheduleUpdate();

	},

	processHoroscopeError: function(error) {
		this.loaded = true;
		this.error = true;
		this.updateDom(this.config.animationSpeed);
		Log.error("Process Horoscope Error: ", error);
	},


	getDom: function() {
		var wrapper = document.createElement("div");
		wrapper.className = "horoscope-wrapper"
		wrapper.style["max-width"] = this.config.maxWidth;

		if (this.config.sign === "") {
			wrapper.innerHTML = "Please set the correct Zodiac <i>sign</i> in the config for module: " + this.name + ".";
			wrapper.className = "dimmed light small";
			return wrapper;
		}

		if (!this.loaded) {
			wrapper.innerHTML = "Aligning Stars...";
			wrapper.className = "dimmed light small";
			return wrapper;
		}

		if (this.error) {
			wrapper.innerHTML = this.name + ": Something went wrong. Please check logs.";
			wrapper.className = "bright light small";
			return wrapper;
		}

		var horoscopeTop = document.createElement("div");
		horoscopeTop.className = "horoscope-top";

		var zodiacIcon = document.createElement("div");
		if (this.config.useTextIcon) {
			zodiacIcon.className = "zodiac-text-icon";
			zodiacIcon.innerHTML = this.sign;
		} else {
			zodiacIcon.className = "zodiac-icon " + this.config.sign;
		}

		var horoscopeTitle = document.createElement("div");
		horoscopeTitle.className = "horoscope-title align-right";

		var zodiacSignText = document.createElement("div");
		zodiacSignText.className = "zodiac-sign-text medium";
		zodiacSignText.innerHTML = this.signText;

		var horoscopeDate = document.createElement("div");
		horoscopeDate.className = "horoscope-date xsmall";
		horoscopeDate.innerHTML = "Horoscope for " + moment(this.horoscopeDate).format("LL");

		horoscopeTitle.appendChild(zodiacSignText);
		horoscopeTitle.appendChild(horoscopeDate);

		horoscopeTop.appendChild(zodiacIcon);
		horoscopeTop.appendChild(horoscopeTitle);

		wrapper.appendChild(horoscopeTop);

		var horoscopeText = document.createElement("div");
		horoscopeText.className = "horoscope-text small";
		horoscopeText.innerHTML = this.horoscopeText;
		wrapper.appendChild(horoscopeText);

		return wrapper;
	},

	scheduleUpdate: function(delay) {
		var nextLoad = this.config.updateInterval;
		if (typeof delay !== "undefined" && delay >= 0) {
			nextLoad = delay;
		}

		var self = this;
		setTimeout(function() {
			self.updateHoroscope();
		}, nextLoad);
	}
});